Tag AWS resources created with Terraform (#1552)

natanlao opened the pull request:
Author (check before every review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label                                   <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes          <sub>or this PR does not need porting</sub> 
- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain            <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox`                             <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox`                                   <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [x] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain                                       <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations                        <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev`                                       <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

Amend Terraform configuration to tag all AWS resources with some
Azul-specific tags (Owner, AzulDeployment, Creator, Name) and some tags that
are specified in [DCP RFC 0006] (project, env, owner, service, Name,
managedBy).

Update documentation in README.md to specify that manually-created resources
should be tagged in a manner consistent with those maintained by Terraform.

If all tags were static, it would make sense to contain tags in the shared
Terraform configuration. Because the content of the tags depends on user config
it is reasonable to maintain helper code in `azul.deployment`.

  [DCP RFC 0006]: https://github.com/HumanCellAtlas/dcp-community/blob/master/rfcs/text/0006-resource-tagging.md

natanlao commented on src/azul/deployment.py:
```diff
@@ -274,3 +278,32 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+        'managedBy': 'terraform'
+    }
+    tags.update(**kwargs)
+    return tags
```
Nice, TIL.

jessebrennan replied:
It looks like you could still do this. Just wondering if there is a reason not to.

natanlao replied:
The `component` tag is set separately because it relies on the `service` tag which in theory could be overriden with `kwargs`

natanlao commented on terraform/cloudwatch.tf.json.template.py:
```diff
@@ -187,6 +193,7 @@
                             f"arn:aws:sns:{aws.region_name}:{aws.account}:cloudwatch-alarms",
                             f"arn:aws:sns:{aws.region_name}:{aws.account}:dcp-events"
                         ],
+                        "tags": get_aws_tags(f"cloudwatch_{queue.replace('.', '_')}_msg_ct"),
```
For consistency with the resource's name (see https://github.com/DataBiosphere/azul/pull/1701/files/c96df34fd536b8c6509d0228435b73b029e04930#diff-ce5e755f913a919a9a8bb34ca5b64346R170)

jessebrennan replied:
This is just the Terraform reference and it doesn't correspond to the actual name of the resource in AWS. The actual resource name is set by the line below it https://github.com/DataBiosphere/azul/pull/1701/files/c96df34fd536b8c6509d0228435b73b029e04930#diff-ce5e755f913a919a9a8bb34ca5b64346R171. I confirmed this by checking the HCA AWS account.

natanlao replied:
Ah, my mistake. I'll fix this.

natanlao commented on src/azul/deployment.py:
```diff
@@ -57,6 +57,10 @@ def apigateway(self):
 
+    @cachedproperty
+    def user(self):
+        return self.sts.get_caller_identity()['UserId']
```
Ah, I see that `AZUL_OWNER` is already documented for this purpose in [environment.py]. Good catch, thanks.

natanlao replied:
I see. I haven't been able to find another way to programmatically retrieve the calling user's email address with the AWS API.

I implemented it this way because the original issue explicitly referenced calling `aws sts get-caller-identity`. `config.owner` is already set in the `Owner` tag. Should I drop the `Creator` tag entirely?

jessebrennan replied:
I kind of think so. We should probably get @hannes-ucsc to weigh in here.

natanlao commented on src/azul/deployment.py:
```diff
@@ -274,3 +278,32 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    tags = {
+        'project': 'dcp',
+        'env': config.deployment_stage,
+        'owner': config.owner,
+        'Owner': config.owner,
```
The purpose, to my recollection, was to comply with both the requirements in the issue description and with the tagging RFC. I'll clarify with Hannes after standup today.

natanlao commented on src/azul/deployment.py:
```diff
@@ -274,3 +278,32 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
     return emit(tf_config) if tf_config is None else emit(_sanitize_tf(tf_config))
+
+
+def get_aws_tags(resource_name: str = '', suffix: str = '', **kwargs) -> dict:
```
Good catch, the call would fail.

natanlao commented on terraform/api_gateway.tf.json.template.py:
```diff
@@ -76,7 +77,8 @@ def for_domain(cls, domain):
                     "name": zone.name,
-                    "private_zone": False
+                    "private_zone": False,
+                    "tags": get_aws_tags(Name=zone.slug)
```
This is a good point. The original thinking was that calling `qualified_resource_name` when setting `Name` would be the rule and not the exception, but after looking at the changes, this is not the case

jessebrennan replied:
I don't feel like there is a satisfying enough reason to mark this comment as "resolved"

natanlao replied:
Sorry, I thought that we discussed this but thinking now I believe I was mistaken. The original rationale was to minimize work for the caller, but looking at the revised guidelines, this assumption is no longer accurate - I'll add these changes to the next revision.

natanlao commented on README.md:
```diff
@@ -373,6 +373,17 @@ supported to use the same zone for both `AZUL_URL_REDIRECT_BASE_DOMAIN_NAME` and
+* `Owner` and `owner`, both set from `AZUL_OWNER`
+* `AzulDeployment` and `env`, both set from `AZUL_DEPLOYMENT_STAGE`
+* `Creator`, set from your AWS STS `UserId` (do `aws sts get-caller-identity`)
+* `Name`, a unique identifier for the hosted zone(s)
```
My understanding was that having duplicate Name tags could be confusing. That being said, I don't do any of this inspection myself, so I can't speak to this personally.

natanlao commented:
Some notes on this latest revision:
- The `name` tag is kept consistent with any existing resource name, even if it's not unique. (See terraform/gitlab/gitlab.tf.json.template.py.) This can cause some irregularity with the `component` tag -- I am not sure if the approach I've taken is the best one.
- The scope of this ticket was expanded to also tag GCP resources, but I found no GCP resources in terraform/ that supported tagging.

jessebrennan commented on terraform/api_gateway.tf.json.template.py:
```diff
@@ -174,7 +174,8 @@ def for_domain(cls, domain):
                             },
-                            "tags": get_aws_tags(f'cloudwatch_{lambda_.name}_group'),
+                            "tags": get_cloud_tags(
+                                name="/aws/apigateway" + config.qualified_resource_name(lambda_.name)),
```
I'm still curious about these special cases. What's the formula or convention for naming tags explicitly?

natanlao replied:
I'm following these guidelines:
* If the resource already has a name assigned to it, use that value. This value should already be qualified.
* If the resource does not already have a name assigned to it, create a unique value in the qualified format (e.g., azul-foo-dev)
* Each name should be unique. (Unless this would deviate from the original resource name)

I've noted general guidelines [here](https://github.com/DataBiosphere/azul/issues/1552#issuecomment-616796430). Should I also copy those guidelines into this PR?

jessebrennan replied:
Yeah I think it would be useful to add this to the docstring of the `get_cloud_tags()` function.

natanlao commented on src/azul/deployment.py:
```diff
@@ -295,38 +296,54 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
-def get_aws_tags(resource_name: str = '', suffix: str = '', **kwargs) -> dict:
+def get_cloud_tags(resource_name: str = '', suffix: str = '', **kwargs) -> dict:
     """
     Returns tags named for AWS resources based on :class:`azul.Config`.
```
Yeah, in theory. That being said, I didn't see any GCP resources set up in Terraform that I could tag...

jessebrennan replied:
```suggestion
    Returns tags named for cloud resources based on :class:`azul.Config`.
```

natanlao commented on src/azul/deployment.py:
```diff
@@ -295,38 +296,54 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    Also renames the "name" tag on any tagged AWS resource to "Name".
     """
-    return {k: v for k, v in tf_config.items() if v}
+    sanitized = {k: v for k, v in tf_config.items() if v}
+    for group in filter(lambda x: isinstance(x, dict), sanitized.get('resource', [])):
+        for resource_name, resource_data in group.items():
+            if resource_name.startswith('aws_'):
+                for service in resource_data.values():
+                    if 'tags' in service and 'name' in service['tags']:
+                        service['tags']['Name'] = service['tags'].pop('name')
```
This makes sense to me. Is this chunk here otherwise clear and well-written? It feels like it could use improvement to me but I am not sure how.

jessebrennan replied:
Yes, otherwise I think it's fine.

natanlao commented:
> For example what happens if resource_name is not specified and neither is the name kwarg passed in explicitly?

This is a good point -- I can add a check for this.

natanlao commented:
Okay, new revision! I've made the following changes:
* `get_cloud_tags` no longer calls `config.qualified_resource_name` directly. Hannes's concern was minimizing effort on the part of the caller – that being said, counting invocations of `get_cloud_tags`, calling `qualified_resource_name` appears now to be the exception (about 1/3 of cases) rather than the rule (the remaining 2/3).
* Updated the docstring of `get_cloud_tags` with guidance as to how the `name` tag is chosen and added some doctests
* Split changes to `_sanitize_tf` into `_fix_name_tag`
* Reviewed choices for the `name` tag across all resources. For the most part, I think those values are reasonable, with the potential exception of what's in `terraform/s3.tf.json.template.py`

natanlao commented:
I've been investigating an issue here where tests will pass but CI will fail on uploading to Codecov and Coveralls. The failures appeared stochastic (as the failure was from a remote rejection of coverage data and localized to only my builds) but persisted over several days. Here is an abridged summary of what I tried:
- Removing `codecov` and `coveralls` from the build individually
- At Hannes's suggestion, pinning `coverage` to 5.x from 4.x.
- Removing the doctests I added (which were working). This removed the failure from the build.
- Re-adding the doctests I removed previously. The build... passed.

So, I'm not sure what happened or why. But it's working now.

hannes-ucsc commented on src/azul/deployment.py:
```diff
@@ -297,5 +301,71 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    >>> assert 'Name' in tags
+    >>> assert 'name' not in tags
+    """
+    for group in filter(lambda x: isinstance(x, dict), tf_config.get('resource', [])):
```
Use the terminology from https://www.terraform.io/docs/configuration/syntax-json.html and be sure you handle [nested blocks](https://www.terraform.io/docs/configuration/syntax-json.html#nested-block-mapping)

hannes-ucsc commented on src/azul/deployment.py:
```diff
@@ -297,5 +301,71 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    >>> resource = {'resource_name': {'tags': {'name': 'foo'}}}
+    >>> fixed = _fix_name_tag({'resource': [{'aws_resource_type': resource}]})
+    >>> tags = fixed['resource'][0]['aws_resource_type']['resource_name']['tags']
+    >>> assert 'Name' in tags
```
I see two conditionals in the implementation so it would make sense to include test cases for which these conditionals evaluate to false.

natanlao commented on src/azul/deployment.py:
```diff
@@ -297,5 +301,71 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+      that name should be used, even if it is not unique. The unique
+      fully-qualified resource name can be generated using
+      :meth:`azul.Config.qualified_resource_name`.
+    * ``project``, set for now to ``dcp``.
```
Okay, I understand. Should this documentation be moved somewhere else (e.g. to the README?), minimized, or removed altogether?

hannes-ucsc replied:
I'd remove it. It is redundant.

natanlao commented on src/azul/deployment.py:
```diff
@@ -297,5 +301,71 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+        'service': config._resource_prefix,
+        'deployment': config.deployment_stage,
+        'owner': config.owner,
+        'creator': aws.user,
```
I was unable to make progress with this, and have removed the `creator` tag entirely. I did find the [`ignore_tags`](https://www.terraform.io/docs/providers/aws/index.html#ignore_tags-configuration-block) property, but it didn't appear to work in the fashion that was required.

hannes-ucsc replied:
I linked to a post about `ignore_changes` but you mention to `ignore_tags`. Did you try the former?

natanlao replied:
Yes -- I took some time today to verify again. I tested with Terraform v0.12.24:

* Using only `ignore_tags` (at the provider level)
* Using only `ignore_changes` (for each resource)
* Using both

None of these approaches were effective; in all of my tests, none of these methods were sufficient to prevent Terraform from trying to update tags on resources that otherwise did not need to be modified.

There is a possibility that I'm doing something wrong -- per the issue that you linked, some people have been able to do this successfully as recently as last year. That being said, there are also a number of people who are unable to reproduce those successes, even though the issue was marked as resolved (hashicorp/terraform#21788).

In any case, it might be worth revisiting this down the line. There is an open Terraform feature request (hashicorp/terraform#20866) that would make implementing this functionality as described much easier.

hannes-ucsc replied:
I'd like to see the version of this PR that uses the technique I linked to and a terminal transcript that proves that it doesn't work.

natanlao replied:
I've tagged you on Slack with a link to a Gist containing relevant transcripts.

natanlao replied:
@hannes-ucsc and I have discussed this at length. The `creator` tag is being removed as the behavior we need (i.e., configuring a tag such that it is only updated if another part of the resource is updated) is not readily achievable. Terraform allows ignoring changes to a tag and updating it only on resource creation or replacement, but this behavior is insufficient for this use case.

hannes-ucsc commented on terraform/api_gateway.tf.json.template.py:
```diff
@@ -73,7 +74,7 @@ def for_domain(cls, domain):
                 zone.slug: {
                     "name": zone.name,
-                    "private_zone": False
+                    "private_zone": False,
```
```suggestion
                    "private_zone": False
```

No trailing commas. See contribution guide.

hannes-ucsc commented on terraform/sqs.tf.json.template.py:
```diff
@@ -29,18 +33,21 @@
                         "message_retention_seconds": 14 * 24 * 60 * 60,
+                        "tags": get_cloud_tags(config.qualified_resource_name('fail')),
                     },
                     config.unqual_tallies_queue_name(fail=True): {
                         "fifo_queue": True,
                         "name": config.tallies_queue_name(fail=True),
                         "message_retention_seconds": 14 * 24 * 60 * 60,
+                        "tags": get_cloud_tags(config.qualified_resource_name('fail', suffix='.fifo')),
```
This looks wrong. It's out of sync with the actual name. This code recently changed so this is probably a merge error.

Anyways, we can already see how brittle this duplication is. Ideally we would want to automate it.

```suggestion
                        "tags": get_cloud_tags("${self.name}"),
```

This doesn't currently work with TF but let's pretend that it does and resolve `${self.name}` in a post-processing step.

Let's also have `${self.name}` be the default such that we can call it as in

```suggestion
                        "tags": get_cloud_tags(),
```

hannes-ucsc replied:
And finally we should add a convenience for the most common case:

```suggestion
                    config.unqual_tallies_queue_name(fail=True): add_tags({
                        "fifo_queue": True,
                        "name": config.tallies_queue_name(fail=True),
                        "message_retention_seconds": 14 * 24 * 60 * 60,
```

hannes-ucsc commented on src/azul/deployment.py:
```diff
@@ -297,5 +298,64 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    return emit(tf_config) if tf_config is None else emit(_fix_name_tag(_sanitize_tf(tf_config)))
+
+
+def get_cloud_tags(resource_name: str, **kwargs) -> Dict[str, str]:
```
```suggestion
def tags(resource_name: str, **kwargs) -> Dict[str, str]:
```

natanlao commented:
After post-standup discussion, changing the implementation such that tags are automatically added to all resources (with exceptions as necessary) in post-processing while retaining the `${self.name}` syntax as recommended. Rationale:

* `${self.name}` closely mirrors existing Terraform syntax.
* Contributors shouldn't have to remember to add tags to new resources.
* If `${self.name}` resolution is left to post-processing, dependent tags (i.e. `component`) must also be resolved in post-processing.
* The implementation should merge with and not overwrite existing tags, preferring existing tags if they exist.

Explicit methods for adding tags (e.g. `add_tags`) should be retained.

natanlao commented:
Build is passing, though for some reason it isn't being reported. [Travis build](https://travis-ci.org/github/DataBiosphere/azul/builds/704113297) for 0f56eda.

jessebrennan commented on src/azul/deployment.py:
```diff
@@ -277,5 +285,149 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    :param kwargs: Provided key-value pairs can be used to specify additional
+                   tags or to override default tags.
+    """
+    component_suffix = config.unqualified_resource_name_or_none(resource_name)[0]
```
I prefer unpacking like this instead of tuple indexing.

```suggestion
    resource_name, stage = config.unqualified_resource_name_or_none(resource_name)
```
or else maybe
```suggestion
    component_suffix, _ = config.unqualified_resource_name_or_none(resource_name)
```

jessebrennan commented on terraform/gitlab/gitlab.tf.json.template.py:
```diff
@@ -532,9 +533,7 @@ def remove_inconsequential_statements(statements: List[JSON]) -> List[JSON]:
-                "tags": {
-                    "Name": "azul-gitlab"
-                }
+                "tags": tags("azul-gitlab", component="azul-gitlab")
```
In my brief test, calling `config.unqualified_resource_name_or_none('azul-gitlab')` returns `(None, None)`.

natanlao replied:
Yes, this is one of the trickier parts of the functionality. You're right that `config.unqualified_resource_name_or_none` (and by extension, `tags()`) will choke on `azul-gitlab` as-is. My workaround is to specify the component manually -- `tags()` won't overwrite an explicitly provided tag, so after post-processing is complete the tags are "correct." Without the workaround, `component` will be `azul-None`.

This definitely isn't preferred behavior, since we want to avoid specifying things manually if we can. Is there a better way to approach this case? I initially thought this was the best approach, but reconsidering it now I don't think it is clear to someone reading the code.

jessebrennan replied:
I see. I don't really have a better idea.

Maybe you could assert that if an unqualified name cannot be parsed, that the component tag is also passed in as an arguement. This might help prevent accidental misuse in the future.

natanlao replied:
I thought about this over the weekend and I agree with that solution -- this case shouldn't fail silently.

jessebrennan commented on src/azul/deployment.py:
```diff
@@ -277,5 +285,149 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    component_suffix = config.unqualified_resource_name_or_none(resource_name)[0]
+    return {
+        'project': 'dcp',
+        'service': config._resource_prefix,
```
Consider renaming `config._resource_prefix` to `config.resource_prefix`

jessebrennan commented on src/azul/deployment.py:
```diff
@@ -277,5 +285,149 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    return emit(tf_config) if tf_config is None else emit(_sanitize_tf(_populate_tags(tf_config)))
+
+
+def tags(resource_name: str, **kwargs) -> Dict[str, str]:
```
It seems like a lot of the arguments you pass in for `resource_name` aren't parsed correctly by `config.unqualified_resource_name_or_none()`. `azul-gitlab` failed for me, but you should test the others too.

jessebrennan commented on src/azul/deployment.py:
```diff
@@ -347,13 +346,17 @@ def _populate_tags(tf_config: JSON) -> JSON:
 
 
-def tags(resource_name: str, **kwargs) -> Dict[str, str]:
+def tags(resource_name: str, **kwargs) -> JSON:
```
I sort of think the previous annotation was more accurate, right? AWS won't accept full JSON for a tag, I think, only a key-value pair.

natanlao replied:
This is something that I went back and forth with for a while. After thinking about it more, I think you're right and I'll change it back.

My rationale was that the `JSON` annotation was accurate for consistency with `def _populate_tags(tf_config: JSON) -> JSON` which itself is consistent with `def _sanitize_tf(tf_config: JSON) -> JSON`. That being said, the linter doesn't like it, because I rely on some `dict` methods through the code. The other option would be `MutableJSON` (alias of `Dict[str, AnyMutableJSON]`), but that doesn't seem to be appropriate here.

hannes-ucsc commented on terraform/gitlab/gitlab.tf.json.template.py:
```diff
@@ -543,17 +542,14 @@ def remove_inconsequential_statements(statements: List[JSON]) -> List[JSON]:
-                "tags": {
-                    "Name": "azul-gitlab"
-                }
+                "tags": tags("azul-gitlab", component="azul-gitlab")
```
It seems like this is screaming for a `gitlab_tags()` wrapper around `tags()` so the component tag doesn't need to be duplicated all over the place.

hannes-ucsc commented on terraform/gitlab/gitlab.tf.json.template.py:
```diff
@@ -569,18 +565,14 @@ def remove_inconsequential_statements(statements: List[JSON]) -> List[JSON]:
-                "tags": {
-                    "Name": f"azul-gitlab-{zone}"
-                }
+                "tags": tags(f"azul-gitlab-{zone}")
```
Why doesn't this specify the component?

hannes-ucsc commented on terraform/gitlab/gitlab.tf.json.template.py:
```diff
@@ -642,6 +632,7 @@ def remove_inconsequential_statements(statements: List[JSON]) -> List[JSON]:
             },
             "gitlab": {
                 "name": "azul-gitlab",
+                "tags": {"component": "azul-gitlab"},
```
Why doesn't this call `tags()`?

natanlao replied:
In cases where the name of a resource doesn't follow the qualified resource name format, the `component` tag can't be inferred, so I specify it manually. For consistency, I'll change this to `tags`

natanlao commented on src/azul/deployment.py:
```diff
@@ -277,5 +285,158 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+        'aws_subnet': None,
+        'aws_vpc': None
+    }
+    name_key = name_key_by_resource.get(resource_type, "name")
```
This is more or less the current behavior, though looking again I can make it more obvious what code needs to be modified and how.

The slight deviation from your description is that `add_tags` will try and infer tagging information and silently add it if possible. If `add_tags` can't do that (i.e., if the resource doesn't have a `name` key that parses nicely) then it will throw an error when Terraform configuration is being generated/validated. Is this the correct behavior?

hannes-ucsc replied:
Yes.

natanlao commented on src/azul/deployment.py:
```diff
@@ -277,5 +285,158 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    """
+    name_key_by_resource = {  # resource_config key containing resource name, or None to skip
+        'aws_acm_certificate': 'domain_name',
+        'aws_acm_certificate_validation': None,
```
It probably would, but I think a blacklist system is necessary to preserve the desired behavior. With a whitelist, it's possible that a developer could specify a resource of a previously-unseen type and forget to configure tagging in this method.

hannes-ucsc replied:
OK

natanlao commented on src/azul/deployment.py:
```diff
@@ -277,5 +285,158 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    return emit(tf_config) if tf_config is None else emit(_sanitize_tf(populate_tags(tf_config)))
+
+
+def tags(resource_name: str, **kwargs) -> Dict[str, str]:
```
PyCharm complains that renaming in this way conflicts with the definition of `tags()`, so I've opted to rename to `user_tags` which I believe fulfills the spirit of your suggestion.

natanlao commented on src/azul/deployment.py:
```diff
@@ -277,5 +285,158 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+        raise RequirementError(f"No {name_key=} in {resource_config=}")
+    else:
+        merged_tags = tags(resource_config[name_key], **resource_config.get("tags", {}))
+        return {**resource_config, "tags": merged_tags}
```
Is this something that I should do in general, or only in this one instance?

natanlao replied:
Post-standup clarification: use whichever is shorter.

natanlao commented:
Huh, I could have sworn that I typed out this comment not an hour ago, but I can't seem to find it now. I'll type it out again anyway...

Progress on this PR after standup discussion with Hannes:
- Chalice creates resources with dashes in them on `make package` (e.g., `servicecachehealth-event`). Fix this by renaming resources using renaming script (update in this PR, since we are changed) and by changing name in post-processing (`scripts/prepare_lambda_deployment.py`) using two passes: first pass to catalog and change resource names in initial definition, and second pass to change references to those names. Can drop the `-event` suffix (e.g. `servicecachehealth` instead).
- Can move script to find taggable resources directly into `azul.terraform` so that it is run on all `make deploy`, can return and address delay (~10s) in future if necessary

natanlao commented on scripts/prepare_lambda_deployment.py:
```diff
@@ -70,6 +77,56 @@ def patch_cloudwatch_resource(resource_type_name, property_name):
+        else:
+            return new_ref
+
+    def _replace_refs(value: literal) -> literal:
```
Is this typed correctly? My intent was to express that the type of `value` will always be the return type.

natanlao commented on src/azul/deployment.py:
```diff
@@ -283,5 +295,174 @@ def _sanitize_tf(tf_config: JSON) -> JSON:
+    for tagged AWS resources.
+
+    >>> from azul.doctests import assert_json
+    >>> assert_json(\
```
Is using `\` to make this span several lines acceptable? Without it, the code would exceed the 120-character limit.

Alternatively, I could extract the argument to `populate_tags` as a variable?

natanlao commented on scripts/prepare_lambda_deployment.py:
```diff
@@ -70,6 +77,56 @@ def patch_cloudwatch_resource(resource_type_name, property_name):
+
+    def _replace_refs(value: literal) -> literal:
+        if isinstance(value, str):
+            for oldref, newref in replace.items():
+                value = value.replace(oldref, newref)
+            return value
+        else:
+            return value
```
Is this fine written as-is or should it be

```suggestion
            for oldref, newref in replace.items():
                value = value.replace(oldref, newref)
        return value
```

The current form keeps the `else` for clarity but the proposed form is more concise.

jessebrennan commented on test/test_tagging.py:
```diff
@@ -0,0 +1,109 @@
+
+    @patch('subprocess.run', new_callable=PropertyMock)
+    def test(self, terraform_mock):
+        # azul.deployment.populate_tags must be tested here because Terraform
+        # is not installed on Travis for unit tests and as such requires mock
```
I appreciate this comment, but I think it's assuming some context that isn't obvious. "must be tested **here**" as opposed to where? Having said that, I don't think that's a question worth answering. It's sufficient to say "We mock Terraform since it's not available on Travis".

I am concerned about the mocked response becoming inaccurate if the Terraform response changes. I think it would be helpful to explain here why that's unlikely to be a problem (because this only depends on the provider, not any of the resources, etc).

hannes-ucsc commented on scripts/prepare_lambda_deployment.py:
```diff
@@ -70,6 +77,56 @@ def patch_cloudwatch_resource(resource_type_name, property_name):
+        'aws_lambda_permission.servicecachehealth-event': 'servicecachehealth',
+    }
+    replace = {k: '.'.join([k.split('.')[0], v]) for k, v in rename.items()}
+    literal = typing.TypeVar('literal', str, int)
```
```suggestion
    T = typing.TypeVar('T', str, int)
```

https://www.python.org/dev/peps/pep-0008/#type-variable-names

If it needs to be hoisted to module scope, so be it.

Independently, "literal" is a bit misleading. "Literal" is a type of AST node, not a type of value. `int` refers to a type, `0` is an integer literal. `1+1` is an expression (another type of AST node) that evaluates to a value of type `int`. "Primitive" would be more accurate but with `T` we avoid the question.

hannes-ucsc commented on scripts/prepare_lambda_deployment.py:
```diff
@@ -70,6 +77,56 @@ def patch_cloudwatch_resource(resource_type_name, property_name):
 
 
+def patch_resource_names(input_json: JSON) -> JSON:
+    # Some Chalice-generated resources have named ending in `-event`. The
```
This should be a docstring.

Might help to add a doctest to illustrate this more.

hannes-ucsc commented on scripts/prepare_lambda_deployment.py:
```diff
@@ -70,6 +77,56 @@ def patch_cloudwatch_resource(resource_type_name, property_name):
     return input_json
 
 
+def patch_resource_names(input_json: JSON) -> JSON:
```
`input` is redundant. Every argument is usually an input. This is a specific type of JSON and the name should reflect that.

natanlao replied:
Should I rename the `input_json` argument in `transform_tf` (https://github.com/DataBiosphere/azul/pull/1701/files#diff-869f5860e107030dfbec5ee431581404R32) as well since I'm working in this file anyway?

hannes-ucsc commented on scripts/prepare_lambda_deployment.py:
```diff
@@ -70,6 +77,56 @@ def patch_cloudwatch_resource(resource_type_name, property_name):
+
+    def _patch_refs(block: JSON) -> JSON:
+        return {
+            k: _patch_refs(v) if isinstance(v, dict) else _replace_refs(v)
```
I don't it is guaranteed that _replace_refs always gets a str or an int. There should be an assert/require somewhere. That may affect the need for a type var.

natanlao replied:
I left this as-is, as `_replace_refs` will return input unaffected if input is not of type `str`. I think the deficiency here was that `T` was poorly defined, so I have redefined it as

    T = typing.TypeVar('T', str, int, float, bool)  # azul.types.PrimitiveJSON

hannes-ucsc replied:
If `block` is `JSON`, then `v` is `AnyJSON`. The conditional only checks for `dict` (incorrectly, BTW, it should be `Mapping`) so that leaves `Sequence` and `PrimitiveJSON` as possible types for the argument to `_replace_refs`. I still think there should be an assert to exclude `Sequence`.

Also, can you think of a way to avoid the brittle coupling between the definition of `T` and PrimitiveJSON?

natanlao replied:
> I still think there should be an assert to exclude `Sequence`.

Ah, I misunderstood previously. I agree with your assessment.

> Also, can you think of a way to avoid the brittle coupling between the definition of `T` and PrimitiveJSON?

My intuition is to unpack `PrimitiveJSON` into the `TypeVar`:

```python
T = typing.TypeVar('T', *PrimitiveJSON.__args__)
```

but PyCharm doesn't like that (`A single constraint is not allowed`). A hacky workaround that PyCharm is okay with is

```python
T = typing.TypeVar('T', *PrimitiveJSON.__args__, typing.NoneType)
```

but that doesn't feel right. Is this the direction I should be looking in?

hannes-ucsc commented on scripts/rename_resources.py:
```diff
@@ -25,7 +25,19 @@
+    'module.chalice_service.aws_cloudwatch_event_target.servicecachehealth-event':
+        'module.chalice_service.aws_cloudwatch_event_target.servicecachehealth',
+    'module.chalice_service.aws_lambda_permission.servicecachehealth-event':
+        'module.chalice_service.aws_lambda_permission.servicecachehealth',
```
This screams for a dict comprehension or some other means of reduction.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,110 @@
+
+class TestTagging(AzulUnitTestCase):
+
+    @patch('subprocess.run', new_callable=PropertyMock)
```
Are you sure you need PropertyMock here? subprocess.run is a method. I think you want a MagicMock here or sth.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,110 @@
+    configure_test_logging()
+
+
+class TestTagging(AzulUnitTestCase):
```
```suggestion
class TestTerraformResourceTags(AzulUnitTestCase):
```

and name module accordingly.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,110 @@
+        }
+        res = populate_tags(tagged_aws_resource)
+        tags = res['resource'][0]['aws_vpc'][0]['name']['tags']
+        self.assertIn('Name', tags)
```
Wouldn't it be more strict and easier to comprehend (albeit more verbose) to assert equality to a complete literal?

natanlao replied:
Tag values can't be hardcoded because they vary depending on stage name, user, etc. I think the code can be clearer and more strict, though. Should I:
* Replicate the logic in azul.deployment._tags in the test code
* Provide a complete "expected" literal and compare only the keys of each dictionary (which would be sufficient and also more strict)
* Something else?

natanlao replied:
I've pushed a commit that uses option #2. I'll wait until you see this before requesting review.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,110 @@
+
+
+if __name__ == '__main__':
+    import unittest
```
Why is this a local import? At this point that module has already been imported. You can verify that by temporarily adding `assert 'unittest' in sys.modules`.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,110 @@
+
+    @patch('subprocess.run', new_callable=PropertyMock)
+    def test(self, terraform_mock):
+        # We mock Terraform since it's not available on Travis. The mocked
```
This comment is misleading. You'd want to mock even if the `terraform` binary were available on Travis.

natanlao replied:
Why would we want to mock even if the `terraform` binary were available on Travis?

Jesse and I discussed this a little bit, and one of the conclusions that we reached is that the schema provided by `terraform providers schema -json` is dependent on the providers used, and not the presence of individual resources, which was one of the reasons why I chose this approach.

Thinking about it now, is the reason we want to mock anyway because we can't guarantee what resources AWS/Terraform will designate as taggable?

hannes-ucsc replied:
Unit tests should have minimal external requirements and be quick. It should be obvious to the informed reader why this is mocking the `terraform` stdout, even without the comment. The reference to Travis is misleading in that it seems to suggest that in other contexts (say, Gitlab, where `terraform` is installed) we would not want to mock.

hannes-ucsc commented on scripts/rename_resources.py:
```diff
@@ -25,9 +25,19 @@
     'aws_sfn_state_machine.cart_item_state_machine': 'aws_sfn_state_machine.cart_item',
     'aws_sfn_state_machine.manifest_state_machine': 'aws_sfn_state_machine.manifest',
-    'null_resource.hmac-secret': 'null_resource.hmac_secret'
+    'null_resource.hmac-secret': 'null_resource.hmac_secret',
```
```suggestion
    'null_resource.hmac-secret': 'null_resource.hmac_secret'
```

hannes-ucsc reviewed (changes requested):
No bump, since I didn't make it fully through.

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -0,0 +1,29 @@
+import functools
```
Same as with typing.

natanlao replied:
Superseded by https://github.com/DataBiosphere/azul/pull/1701#discussion_r490632606

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -0,0 +1,29 @@
+
+
+@functools.lru_cache
+def taggable_resources() -> Sequence[str]:
```
This module is a bit anemic and I don't really see anything being added to it anytime soon. This should be defined in `deployment.py`.

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -0,0 +1,29 @@
+)
+
+
+@functools.lru_cache
```
I dislike use of `lru_cache` when there are no parameters that warrant the dict caching employed by `lru_cache`

Convert to method of a singleton class (see AWS in `deployment`) and apply `@cached_property` instead.

Class and instance names should be `Terraform` and `terraform` respectively.

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -0,0 +1,29 @@
+    require(schema['format_version'] == '0.1')
+    resources = itertools.chain.from_iterable(
+        schema['provider_schemas'][provider]['resource_schemas'].items()
+        for provider in schema['provider_schemas'])
```
) on new line

hannes-ucsc commented on src/azul/terraform.py:
```diff
@@ -0,0 +1,29 @@
+    resources = itertools.chain.from_iterable(
+        schema['provider_schemas'][provider]['resource_schemas'].items()
+        for provider in schema['provider_schemas'])
+    return [resource_type for resource_type, resource in resources
```
wrap all and hanging indent

hannes-ucsc commented on test/test_doctests.py:
```diff
@@ -65,9 +67,12 @@ def load_tests(_loader, tests, _ignore):
                    load_module(root + '/scripts/envhook.py', 'envhook'),
                    load_module(root + '/scripts/export_environment.py', 'export_environment'),
-                   load_module(root + '/scripts/check_branch.py', 'check_branch'),
+                   load_module(root + '/scripts/prepare_lambda_deployment.py',
+                               'prepare_lambda_deployment'),
```
Dislike how this line sticks out. It must be longer than 80 but If you have it in you can you refactor this so the line reads

```suggestion
                   load_script('prepare_lambda_deployment'),
```

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,222 @@
+        """
+        try:
+            self.assertDictEqualPermissive(expected, actual)
+        except AssertionError:
```
This is weird. 

I was under the impression that we agreed that the helper method returns bool and contains the meat of the functionality as well as the doctests. Did we misunderstand each other or did you intentionally deviate from the agreement for a particular reason? If the latter, you should pro-actively draw my attention to that and outline your reasoning.

natanlao replied:
I think we misunderstood each other. I neglected to record the exact details of our conversation in my notes, but from what I remember:

* We agreed that `self.assert*` should be used to make it easier to identify deviations between the expected and observed output.
* We also agreed that writing out `AssertionError`s in doctests would be verbose.

My implementation was largely guided by these two points.

hannes-ucsc replied:
Yes, those two bulleted items were part of our agreement. I also believe that we agreed to write a worker method that returns a boolean and a wrapper method that calls that worker method and raises an assertion error if the return value is False. The current implementation does the inverse, so to speak. The "pretty diff" directive gets in the way of the "boolean worker" directive which is probably why you went that route.

Let's tweak this a little bit. Instead of True/False, have the worker return the path of the discrepancy or None if there is no significant difference. Include the path in the assertion error that the wrapper raises. I hope that's going to simplify things.

hannes-ucsc commented:
If you feel that I am contradicting my own guidance from earlier review cycles, don't hesitate to point that out.

natanlao commented:
Please note 887eb08, which is new.

hannes-ucsc commented on src/azul/deployment.py:
```diff
@@ -280,6 +281,32 @@ def resource(self, *args, **kwargs):
 aws = AWS()
 
 
+class Terraform(object):
```
```suggestion
class Terraform:
```

natanlao replied:
I didn't realize that all classes are new-style by default in Python 3. Should this be added to the contributing guide?

hannes-ucsc replied:
I don't think so. I'd file that under common sense.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,241 @@
+        path = self.assertDictEqualPermissive(expected, actual)
+        return self.assertIsNone(path, f'Discrepancy in path: {path}')
+
+    def assertDictEqualPermissive(self, expected, actual, path: str = '') -> Optional[str]:
```
```suggestion
    def assertDictEqualPermissive(self, expected, actual, path: Tuple[str] = ()) -> Optional[str]:
```

or

```suggestion
    def assertDictEqualPermissive(self, expected, actual, *path: str) -> Optional[str]:
```

Whichever is more convenient.

hannes-ucsc replied:
This method compares. It does not assert. Hence it should not have "assert" in its name.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,241 @@
+
+class TestTerraformResourceTags(AzulUnitTestCase):
+
+    def permissive_compare(self, expected, actual):
```
This method asserts. So it should have "assert" in its name.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,241 @@
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 'abc'}]}
+        ... )
+        ...
+        Traceback (most recent call last):
```
These traceback literals is what we wanted to avoid in the first place. The doctests should be on the method that does the comparison, not the one that asserts.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,235 @@
+                    diff = self.permissive_compare(expected[k], actual[k], *path, str(k))
+                    if diff is not None:
+                        return diff
+                    else:
+                        pass
```
```suggestion
```

and everywhere else

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,235 @@
+        path = self.permissive_compare(expected, actual)
+        return self.assertIsNone(path, f'Discrepancy in path: {path}')
+
+    def permissive_compare(self, expected, actual, *path: str) -> Optional[str]:
```
type hints for arguments

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,235 @@
+        Returns None if dictionaries match, or path of discrepancy as a
+        string otherwise.
+
+        >>> t = TestTerraformResourceTags()
```
These doctests don't seem to cover all the branches:

![image](https://user-images.githubusercontent.com/5143256/94844161-aae3bf80-03d2-11eb-9175-fe30bd8ea0b3.png)

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,235 @@
+                        return diff
+                    else:
+                        pass
+                elif isinstance(v, list):
```
I think this would benefit being turned inside out so that the type check occurs on the arguments, instead of the values of the arguments. It should make it more generally useful may also make it shorter. Since I am guessing here, be sure to make this change in a separate fixup commit so we can revert it easily. That fixup should also be the last one on the branch.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,235 @@
+        """
+        # in case actual is a superset of expected
+        if len(actual) != len(expected):
+            return 'object length mismatch'
```
This seems weird. This should return the path, no? Maybe the function should return the path *and* a string explaining the difference. But just the path is fine; at the root that would be the empty tuple.

hannes-ucsc commented:
Feel free to squash out the fixups before adding more fixups. I forgot to do that.

hannes-ucsc reviewed (changes requested):
As covered in the C-guide, I prefer not to have more than one fixup on the same base commit between review cycles.

natanlao commented:
@hannes-ucsc 

> As covered in the C-guide, I prefer not to have more than one fixup on the same base commit between review cycles.

My apologies, I misinterpreted [your comment](https://github.com/DataBiosphere/azul/pull/1701#discussion_r498414718). Should I have differed the commit message for one of the fixup commits to differentiate it?

natanlao commented on test/test_tagging.py:
```diff
@@ -0,0 +1,230 @@
+    def permissive_compare(self,
+                           expected: Any,
+                           actual: Any,
+                           *path: Hashable) -> Optional[Tuple]:
```
Would `Hashable` not be appropriate, since dicts can have non-str keys? Or should I cast all dict keys to `str` and hint `path` as `str`?

hannes-ucsc replied:
A previous implementation concatenated the keys into a string with dots in between and now you are arguing for the most general type for the keys even though your tests don't even exercise that case?

Please make the keys `str` and the two main arguments `AnyJSON`.

natanlao commented on test/test_tagging.py:
```diff
@@ -0,0 +1,230 @@
+
+class TestTerraformResourceTags(AzulUnitTestCase):
+
+    def assertDictEqualPermissive(self, expected: Any, actual: Any):
```
You're correct, I mistakenly applied the hints to permissive_compare to this method.

In this case, should TestTerraformResourceTags.permissive_compare retain explicit `Any` hints, or should I omit them since it is implied?

hannes-ucsc replied:
My original comment already answers that:

> The argument types should be more restrictive.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,232 @@
+        ('qaz', '[0]', 'qux', '123')
+        """
+        if isinstance(actual, list) and isinstance(expected, list):
+            for i, v in enumerate(zip(expected, actual)):
```
The use of `zip` causes extra elements in actual to be ignored.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,232 @@
+
+class TestTerraformResourceTags(AzulUnitTestCase):
+
+    def assertDictEqualPermissive(self, expected: Mapping, actual: Mapping):
```
Should match the helper method's signature, obviously.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,232 @@
+                if diff is not None:
+                    return diff
+        elif isinstance(actual, dict) and isinstance(expected, dict):
+            for k, v in expected.items():
```
Doesn't take extra elements in `actual` into account.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,232 @@
+
+        >>> t = TestTerraformResourceTags()
+        >>> t.permissive_compare(
+        ...     {'foo': 'bar', 'qaz': [{'qux': 123, 456: 789}]},
```
`int` keys are invalid JSON.

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,232 @@
+        """
+        if isinstance(actual, list) and isinstance(expected, list):
+            for i, v in enumerate(zip(expected, actual)):
+                diff = self.permissive_compare(*v, *path, f'[{i}]')
```
We can avoid inventing syntax by relaxing the type constraint on keys to `Union[str, int]`

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,232 @@
+                try:
+                    diff = self.permissive_compare(v, actual[k], *path, str(k))
+                except KeyError:
+                    return path + (str(k),)
```
```suggestion
                    return *path, k
```

hannes-ucsc commented on test/test_tagging.py:
```diff
@@ -0,0 +1,232 @@
+                           actual: AnyJSON,
+                           *path: str) -> Optional[Tuple[str]]:
+        """
+        Recursive dictionary comparison. Skips comparison for value in
```
Out of date with the actual implementation.

hannes-ucsc reviewed (commented):
Some of the adjustments in the last cycle were pretty sloppy. To cut this short, I'll make the necessary changes myself.

hannes-ucsc commented:
> My apologies, I misinterpreted your comment. Should I have altered the commit message for one of the fixup commits to differentiate them, or made one of the fixup commits against a different base commit?

I apologize. I had forgotten that I asked for that. A note in the commit message body of both commits might have helped but I'm not going to pretend that I would have noticed that either.

hannes-ucsc commented on src/azul/deployment.py:
```diff
@@ -268,6 +281,32 @@ def resource(self, *args, **kwargs):
+        schema = subprocess.run(['terraform', 'providers', 'schema', '-json'],
+                                cwd=os.path.join(os.environ['project_root'], 'terraform'),
+                                check=True,
+                                capture_output=True,
```
We shouldn't capture stderr.

hannes-ucsc reviewed (changes requested):
Sandbox build failed. Doesn't look like something my changes caused.

natanlao commented:
Sandbox testing revealed a chicken-and-egg relationship in this code; `make package deploy` (which generates .tf.json files) indirectly calls `terraform providers schema -json`, which requires presence of said .tf.json files. After standup discussion with Hannes, fix for this issue is:

1. Add two targets to `terraform/Makefile`:
  * `schema`, which generates a schema file that will be checked in to the repository. The file contains a JSON object with two keys, one with the output of `terraform -version` and the other with the output of `terraform providers schema -json`. It must depend on the `init` target.
  * `check_schema`, which runs `terraform -version` and compares it to the contents of the aforementioned schema file. It should depend on no other targets (except maybe `check_terraform`?). If this target fails, it should print instructions on how to rectify the failure.
2. Change tagging logic to rely on the checked-in schema file for determining what resources are taggable instead of running `terraform providers schema -json`.
3. Add `make -C terraform check_schema` to the GitLab build.

natanlao commented:
@hannes-ucsc It occurs to me that the approach we've chosen will require `make schema` to be run every time a new Terraform version is released. Here is example output from `terraform -version`:

```
Terraform v0.12.24
+ provider.aws v2.70.0
+ provider.google v2.20.3
+ provider.null v2.1.2
+ provider.template v2.1.2

Your version of Terraform is out of date! The latest version
is 0.13.4. You can update by downloading from https://www.terraform.io/downloads.html
```

hannes-ucsc commented:
We should discard the footer. I think splitting on `\n\n` might be an option.

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,64 @@
+schema_file = os.path.join(terraform_dir, '_schema.json')
+
+
+def terraform_cmd(*args: str) -> str:
```
In the other script a similar method is called `terraform_state` so this would imply that this method should be called just `terraform`

natanlao replied:
Standup: Hannes says to rename the functions to `run_terraform` and `run_terraform_state`.

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,64 @@
+
+def generate_schema_json() -> None:
+    schema = terraform_cmd('providers', 'schema', '-json')
+    versions, footer = terraform_cmd('-version').split('\n\n')
```
A comment should explain why footer is ignored

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,64 @@
+        json.dump({
+            'versions': versions,
+            'schema': json.loads(schema)
+        }, f)
```
wouldn't hurt to add `indent=4` this so we can actually diff this file.

hannes-ucsc replied:
Also not so hot about the wrapping style. I would extract the JSON literal into a variable.

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def check_schema_json() -> None:
+    versions, footer = terraform_cmd('-version').split('\n\n')
```
Duplicates a line in previous method.

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,64 @@
+def check_schema_json() -> None:
+    versions, footer = terraform_cmd('-version').split('\n\n')
+    with open(schema_file, 'r') as f:
+        versions_cached = json.load(f)['versions']
```
This isn't exactly caching. A cache usually isn't perturbed by being deleted.

natanlao replied:
Hannes prefers "tracked" or "committed" in that order

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,64 @@
+    if versions == versions_cached:
+        return
+    else:
+        raise RuntimeError("Cached Terraform schema is out of date. Run "
```
Ditto

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,64 @@
+        return
+    else:
+        raise RuntimeError("Cached Terraform schema is out of date. Run "
+                           "`make -C terraform schema` and commit terraform/_schema.json.")
```
Hard-codes file name

hannes-ucsc commented on src/azul/deployment.py:
```diff
@@ -285,13 +284,11 @@ class Terraform:
-        schema = json.loads(schema.stdout)
+        schema_path = os.path.join(os.environ['project_root'],
+                                   'terraform',
+                                   '_schema.json')
```
hard-coded file name

natanlao replied:
Is there a preferred way to extract this? Potential approaches I can think of:

* Extract the path into a variable in `environment`
* Define the path as a constant in `src.azul.deployment` and reference that path in `scripts/terraform_schema.py`
* Require path to be provided as an argument in `scripts/terraform_schema.py` (I chose this initially, but I can't think of a simple way to provide the same path to `src.azul.deployment` with this approach)

natanlao replied:
Hannes prefers approach 1.

hannes-ucsc commented on src/azul/deployment.py:
```diff
@@ -285,13 +284,11 @@ class Terraform:
+        schema_path = os.path.join(os.environ['project_root'],
+                                   'terraform',
+                                   '_schema.json')
+        with open(schema_path) as f:
```
Consider extracting reading and writing the file into functions.

natanlao replied:
After implementing the other changes you suggested in your review, I don't think this extraction is necessary, so I'm proceeding without it. I'm happy to revisit it if you disagree

natanlao replied:
Hannes ok'd this during standup.

hannes-ucsc commented on environment.py:
```diff
@@ -418,5 +418,8 @@ def env() -> Mapping[str, Optional[str]]:
+        'azul_velocity_zenhub_url': None,
+
+        # Path to tracked Terraform schema used to generate *.tf.json files.
+        'AZUL_TRACKED_SCHEMA_PATH': '{project_root}/terraform/_schema.json'
```
There is no point in making this configurable. It should be hard-coded in Config. At least that was my intention when we discussed this at stand-up.

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,74 @@
+    return versions
+
+
+def generate_schema_json(path: str) -> None:
```
```suggestion
def generate_schema_json() -> None:
```

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,74 @@
+                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('command', choices=commands)
+    arguments = parser.parse_args()
+    commands[arguments.command](config.tracked_schema_path)
```
… since this is the only call site and I don't think anyone would want to call the function with a different value.

hannes-ucsc commented on src/azul/__init__.py:
```diff
@@ -782,6 +788,10 @@ def dynamo_object_version_table_name(self) -> str:
 
+    @property
+    def tracked_schema_path(self) -> str:
+        return os.path.join(self.project_root, 'terraform/_schema.json')
```
We prefer `pathlib.Path`.

natanlao replied:
To clarify, does this mean that I should return `tracked_schema_path` as pathlib.Path, I should use pathlib.Path instead of os.path.join, or both?

natanlao replied:
(The first one.)

hannes-ucsc replied:
We are moving to pathlib. When modifying an existing module the choice of whether to update the module to pathlib or to post-pone that and stick with os.path withing the module is at the discretion of the author. The decision should be guided by the urgency of the issue at hand vs the complexity of the update.

hannes-ucsc commented on src/azul/deployment.py:
```diff
@@ -296,6 +307,27 @@ def resource(self, *args, **kwargs):
+    @cached_property
+    def taggable_resources(self) -> Sequence[str]:
+        with open(config.tracked_schema_path) as f:
+            schema = json.load(f)['schema']
```
Don't like how the knowledge about the internal format of that file is distributed over the codebase. Consolidate the reading and writing of the file in a class, probably this one.

hannes-ucsc commented on scripts/rename_resources.py:
```diff
@@ -29,6 +29,16 @@
+renamed.update({name: name.split('-')[0] for name in rename_chalice})
+
 
 def terraform_state(*args):
```
```suggestion
def run_terraform_state(*args):
```

hannes-ucsc reviewed (changes requested):
Remove merge conflict markers,

natanlao commented:
My mistake, sorry about that.

hannes-ucsc commented on scripts/prepare_lambda_deployment.py:
```diff
@@ -66,6 +81,101 @@ def patch_cloudwatch_resource(resource_type_name, property_name):
+        'aws_cloudwatch_event_rule.servicecachehealth-event': 'servicecachehealth',
+        'aws_cloudwatch_event_target.servicecachehealth-event': 'servicecachehealth',
+        'aws_lambda_permission.servicecachehealth-event': 'servicecachehealth',
+    }
```
While it's OK to hard code the list of resources in rename_resources.py, hard-coding it here seems very brittle. There must be a less pedestrian way to code this.

natanlao replied:
My mistake. I thought the purpose of your [fixup](https://github.com/DataBiosphere/azul/pull/1701/commits/63e71d16903a168f06e2c01a651248655b464148#diff-b6226abb933cef28636f31a369c1d7f96b90c211cb9c29f30f3455a95b55ea06R40) was to address this comment.

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,75 @@
+
+def generate_schema_json() -> None:
+    schema = run_terraform('providers', 'schema', '-json')
+    tracked = {
```
https://github.com/DataBiosphere/azul/pull/1701#discussion_r513836024

Please centralize reading and writing and parsing of the schema file in the Terraform class.

hannes-ucsc reviewed (changes requested):
I'll push some fixups, too.

hannes-ucsc commented on scripts/terraform_schema.py:
```diff
@@ -0,0 +1,42 @@
+
+
+def check_schema_json() -> None:
+    if terraform.versions() == terraform.tracked_schema['versions']:
```
Still leaky. It leaks the fact that the schema is stored as a dictionary and that one of the keys in that dictionary is 'versions`.

natanlao replied:
Hannes says versions and schema subkeys should be exposed through the Terraform class as properties, and that there should be a method that takes those two properties and writes them to a file.

hannes-ucsc commented:
For demo, show labelled resources.
